pkg/streamingpromql/functions/common.go

```diff
@@ -43,3 +46,7 @@ func FloatTransformationDropHistogramsFunc(transform func(f float64) float64) In
 func Passthrough(seriesData types.InstantVectorSeriesData, _ *pooling.LimitingPool) (types.InstantVectorSeriesData, error) {
 	return seriesData, nil
 }
+
+// RangeVectorStepFunction is a function to operate on a range vector step.
+// floatBuffer and histogramBuffer will contain the values for the step range.
+type RangeVectorStepFunction func(step types.RangeVectorStepData, rangeSeconds float64, floatBuffer *types.FPointRingBuffer, histogramBuffer *types.HPointRingBuffer) (*promql.FPoint, *promql.HPoint, error)
```

Comment:
I'd prefer not to return pointers for `FPoint` and `HPoint` here - perhaps we could return `bool`s to indicate their presence instead of checking the points for `nil`?

Reply:
One more idea: rather than returning a `FPoint` or `HPoint`, what if this returned `(float64, bool)` or `*FloatHistogram`? Then the implementation of the function doesn't need to think about the output sample timestamp.

Reply:
What is the opposition to returning a pointer here?

I don't like that we have to have a value for the float64 that we only check if a bool tells us to.

The RangeVectorStepFunction needs the `step` so it already understands the timestamp.

We only create the FPoint or HPoint if it is needed, so returning nil should be minimal overhead.

Reply:
>What is the opposition to returning a pointer here?

It adds significantly to garbage collection: we'd be creating a new pointer for every single point emitted by a function over a range vector, only to immediately drop it on the floor once we store the values in the `[]FPoint` / `[]HPoint` slice we've already allocated for the `InstantVectorSeriesData`.

>The RangeVectorStepFunction needs the `step` so it already understands the timestamp.

Agree - but if we just return the `float64` or `*FloatHistogram`, then the function-specific implementation can be even simpler, as most of the implementations then become "call `floatBuffer.UnsafePoints(step.RangeEnd)`, compute result over slices, return result".

Reply:
If you're curious about the impact of the pointers, I'd suggest comparing the benchmarks for `rate()` before and after the changes in this PR - I would expect a significant impact.

Reply:
Okay, makes sense, thanks.